vm/src/objint.rs

```diff
@@ -19,8 +20,70 @@ fn get_value(obj: PyObjectRef) -> i32 {
     }
 }
 
+fn int_add(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
+    arg_check!(
+        vm,
+        args,
+        required = [(i, Some(vm.ctx.int_type())), (i2, None)]
+    );
+    if objtype::isinstance(i2.clone(), vm.ctx.int_type()) {
```

Comment:
I think the body of this function should be something like:

```
Ok(vm.ctx.get_int(get_value(i.clone()) + get_value(getattribute(vm, i2, "__int__")))
```

Ie. we should try to coerce the right argument to an int rather enumerate all possible types. (Need to look carefully at how that works because in ```2 + 2.5``` we should coerce the left to a float instead.)

Reply:
I agree that coercion is a little weird. There does not seem to be one silver bullet way, but it is some sort of compromise. Sometimes, in case of bool and str, you can simply call a method on the object to do the type conversion. In other cases, the constructor or operator call takes care of the conversion inside, this is the case for int and float for example.

Reply:
Maybe someday we can move the __add__ logic to the object type instead of a special case for each first type. But we can see that later on. First we need to have `type(2 + 2) is int` right.